### DISCUSSION TITLE
How do you type decorators for `getServerSideProps`?

### DISCUSSION THREAD

**Discussion by GabenGar** [vercel/next.js#37287] [Help]
In a decorator like this:
```typescript
import type { ParsedUrlQuery } from "querystring";
import type {
  GetServerSideProps
} from "next";
// some baseline props expected to be returned by all pages
interface BasePageProps extends Record<string, unknown> {
  locale: string;
  defaultLocale: string
}
// various configuration options
interface IGetServerSidePropsOptions {}

export function getServerSidePropsDecorator<
  Props extends BasePageProps,
  Params extends ParsedUrlQuery
>(
  options: IGetServerSidePropsOptions,
  callback: GetServerSideProps<Props, Params>
): typeof callback {
  const {} = options;

  return async (...args: Parameters<typeof callback>) => {
    const [context] = args;
    const { locale, defaultLocale } = context;

    try {
      const result = await Promise.resolve(callback(context));
      if (result.notFound || result.redirect) {
        return result;
      }
      
      const newResult = { locale, defaultLocale, ...result.props };

      return newResult;
    } catch (error) {
      const isProperError = error instanceof Error;

      // rethrow unknown errors
      if (!isProperError) {
        throw error;
      }
      console.error(error);

      return {
        redirect: {
          destination: "/500",
          statusCode: 302,
        }
      };
    }
  };
}
```
The type of `result` cannot be narrowed down to either `redirect`, `notFound` or `props` through conditional checks.
And they all point to different errors:
- `result.notFound` - Property 'notFound' does not exist on type '{ redirect: Redirect; }'.ts(2339)
- `result.redirect` - Property 'redirect' does not exist on type '{ notFound: true; }'.ts(2339)
- `result.props` - Property 'props' does not exist on type '{ redirect: Redirect; }'.ts(2339)

Am I missing something or did I stumble upon a bug in types?

**Comment by icyJoseph** [comment]
Try something like this:


```ts
import type {ParsedUrlQuery} from 'querystring';

import type {GetServerSideProps} from 'next';
// some baseline props expected to be returned by all pages
interface BasePageProps extends Record<string, unknown> {
    locale: string;
    defaultLocale: string;
}
// various configuration options
interface IGetServerSidePropsOptions {
    foo: string;
}

export function getServerSidePropsDecorator<Props extends BasePageProps, Params extends ParsedUrlQuery>(
    options: IGetServerSidePropsOptions,
    callback: GetServerSideProps<Props, Params>
): typeof callback {
    const {foo} = options;

    const decorated = async (...args: Parameters<typeof callback>) => {
        const [context] = args;
        const {locale, defaultLocale} = context;

        try {
            const result = await callback(context);

            if ('notFound' in result || 'redirect' in result) {
                return result;
            }

            if ('props' in result) {
                const data = await result.props;
                /* data's locale, defaultLocale override these two
                Maybe do this instead:

                const {locale:_, defaultLocale: __, ...rest} = await result.props;
                return {locale, defaultLocale, ...rest}

                or put locale, and defaultLocale at the end (after data)
                */

                return {props: {locale, defaultLocale, ...data}};
            }
            throw new Error('Invalid result, missing props, redirect, or notFound');
        } catch (error) {
            const isProperError = error instanceof Error;

            // rethrow unknown errors
            if (!isProperError) {
                throw error;
            }
            console.error(error);

            return {
                redirect: {
                    destination: '/500',
                    statusCode: 302,
                },
            } as const;
        }
    };

    return decorated;
}
```

I believe, you missed some TypeScript steps, and you are exposing yourself to prop overriding in `locale`, `defaultLocale`, perhaps by design.

Generally, to narrow down a discriminated union, you cannot do property read outs. TypeScript wants you to do a sort of `does it have it?` approach, instead of `I read it, and it was there`, because the later fails when the value is `undefined`, but it is proper, as opposed to not being there at all. 

This page in the TypeScript docs, https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing, describes how to use the `in` operator to narrow down types. Though, many will point out that `in` in JavaScript checks the entire prototype chain of the target object.

Last, but not least, the `redirect` in the catch block, needs an `as const` because otherwise, it might be seen as `redirect` being `undefined` in the upper `try` block, but what we really mean is that these are discriminated (do not overlap).

**Reply by GabenGar** [reply]
> I believe, you missed some TypeScript steps, and you are exposing yourself to prop overriding in locale, defaultLocale, perhaps by design.

Yes I have to pass a bunch of boilerplate to the props so it could be used by all pages during the server render. So I just want to make some of them optional for the callback but required for the decorated function.

I still can't quite get without Typescript erroring out with .
The signature looks like this (without options):
```typescript
import type { ParsedUrlQuery } from "querystring";
import type { GetServerSideProps } from "next";

interface IBasePageProps {
  locale: string
}
type ISlimProps<Props extends IBasePageProps = IBasePageProps> = Omit<Props, "locale"> & {
  locale?: string
}

function getServerSidePropsDecorator<
  Props extends IBasePageProps,
  Params extends ParsedUrlQuery
>(
  callback: GetServerSideProps<ISlimProps<Props>, Params>
): GetServerSideProps<Props, Params>
```
The errors pops up at the stage of transforming the result with `ISlimProps` into `Props`. 
But if I do `const newProps: { props: Props } = {locale, ..slimProps}` then it throws a bunch of nested errors about instantiation.
Is type assertion the only way in situations like these?

**Reply by icyJoseph** [reply]
> But if I do const newProps: { props: Props } = {locale, ..slimProps} then it throws a bunch of nested errors about instantiation.

Is it complaining about `redirect` and friends? then yeah, the result of GSSP is a union, of three types, and you can't, in the eyes of TS, return ISlimProps as a representative of the entire union. I'd need to see more of the implementation to feel sure about what I am claiming though.

**Reply by GabenGar** [reply]
No, it's complaining about the shape of props I think, since I do filter out non-props results beforehand with `if (!("props" in result))`.
Implementation currently looks like this: https://github.com/GabenGar/next-js-site/blob/381cfd1fc147f29af14a0c801da18d1716bac137/src/server/requests/get-server-side-props.ts#L14
As far as actual workings, it does what I wanted, i.e. typescript accepts slimmer props from the callback, but complains if they don't fulfill page-specific props and the page gets the full props in the end. Even with the prop error ignored.

**Reply by icyJoseph** [reply]
I'd go with something like:

```ts
/**
 * Props shared across all pages.
 */
export type BasePageProps<OwnProps = unknown> = OwnProps &
  SSRConfig & {
    /**
     * This info has to be passed to generate links server-side.
     */
    localeInfo: {
      locale: ISOLangString;
      defaultLocale: ISOLangString;
    };
    errors?: Array<string>;
    schemaValidationErrors?: ISchemaValidationError[];
  };

export type ISlimProps<OwnProps> = OwnProps &
  Omit<BasePageProps, "localeInfo" | "_nextI18Next"> & {
    localeInfo?: ILocaleInfo;
    _nextI18Next?: SSRConfig[keyof SSRConfig];
  };
```

And then the wrapper:

```ts
export function createServerSideProps<
  OwnProps,
  Params extends ParsedUrlQuery = ParsedUrlQuery
>(
  options: IPageOptions,
  callback: GetServerSideProps<ISlimProps<OwnProps>, Params>
): GetServerSideProps<BasePageProps<OwnProps>, Params> {
 
// later on
  const newProps = {
```

And the typo in `all.tsx`'s getServerSideProps:

```ts
 { extraLangNamespaces: ["frontend-mentor"] },
```

**Reply by GabenGar** [reply]
Is this something not doable with just interfaces? One of the reasons I don't like `type` for describing object shapes is typescript presents them as literals in the errors and tool-tips and thus it takes a while to even guess what types it means in the first place.

**Reply by GabenGar** [reply]
Your approach definitely works better by also cutting down on imports on the pages. 
But another problem is how to type the decorator with the callback being optional? 
I thought something like this within the decorated function would suffice:
```
if (!callback) {
  return {
    props: {
      ...localization,
      localeInfo,
    },
  };
}
```
This should meet the `BasePageProps<unknown>` requirement, but it doesn't.
